taichi/program/state_flow_graph.cpp

```diff
@@ -427,9 +514,212 @@ std::string StateFlowGraph::dump_dot(
         }
       }
     }
+    TI_WARN_IF(
+        visited.size() > nodes_.size(),
+        "Visited more nodes than what we actually have. The graph may be "
+        "malformed.");
   }
   ss << "}\n";  // closes "dirgraph {"
   return ss.str();
 }
 
+void StateFlowGraph::topo_sort_nodes() {
+  std::deque<std::unique_ptr<Node>> queue;
+  std::vector<std::unique_ptr<Node>> new_nodes;
+  std::vector<int> degrees_in(nodes_.size());
+
+  reid_nodes();
+
+  for (auto &node : nodes_) {
+    int degree_in = 0;
+    for (auto &inputs : node->input_edges) {
+      degree_in += (int)inputs.second.size();
+    }
+    degrees_in[node->node_id] = degree_in;
+  }
+
+  queue.emplace_back(std::move(nodes_[0]));
+
+  while (!queue.empty()) {
+    auto head = std::move(queue.front());
+    queue.pop_front();
+
+    // Delete the node and update degrees_in
+    for (auto &output_edge : head->output_edges) {
+      for (auto &e : output_edge.second) {
+        auto dest = e->node_id;
+        degrees_in[dest]--;
+        TI_ASSERT(degrees_in[dest] >= 0);
+        if (degrees_in[dest] == 0) {
+          queue.push_back(std::move(nodes_[dest]));
+        }
+      }
+    }
+
+    new_nodes.emplace_back(std::move(head));
+  }
+
+  TI_ASSERT(new_nodes.size() == nodes_.size());
+  nodes_ = std::move(new_nodes);
+  reid_nodes();
+}
+
+void StateFlowGraph::reid_nodes() {
+  for (int i = 0; i < nodes_.size(); i++) {
+    nodes_[i]->node_id = i;
+  }
+  TI_ASSERT(initial_node_->node_id == 0);
+}
+
+void StateFlowGraph::replace_reference(StateFlowGraph::Node *node_a,
+                                       StateFlowGraph::Node *node_b) {
+  // replace all edges to node A with new ones to node B
+  for (auto &edges : node_a->output_edges) {
+    // Find all nodes C that points to A
+    for (auto &node_c : edges.second) {
+      // Replace reference to A with B
+      if (node_c->input_edges[edges.first].find(node_a) !=
+          node_c->input_edges[edges.first].end()) {
+        node_c->input_edges[edges.first].erase(node_a);
+
+        node_c->input_edges[edges.first].insert(node_b);
+        node_b->output_edges[edges.first].insert(node_c);
+      }
+    }
+  }
+  node_a->output_edges.clear();
+}
```

Comment:
If I understand correctly what this function does, it should also replace all edge *from* node A with new ones from node B?

Reply:
Should be fixed in #1872.

Reply:
Thanks!